activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -976,8 +976,10 @@ def build_left_outer_joins(manager, outer_joins)
           case join
           when Hash, Symbol, Array
             :association_join
+          when ActiveRecord::Associations::JoinDependency
+            :stashed_join
           else
-            raise ArgumentError, "only Hash, Symbol and Array are allowed"
+            raise ArgumentError, "only Hash, Symbol, Array and JoinDependency are allowed"
```

Comment:
JoinDependency is not public API and we don't want users to be using it in their applications so we should not talk about it here.